### PR TITLE
Docker utils

### PR DESCRIPTION
**Pull request by raonismaneoto**
This pr add a module that was placed on the server. This module helps the interaction between the system and the docker cli. Thanks to @wesleymonte @emanueljoivo for this.

### CODE REVIEW

File: utils/docker_utils.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+func RemoveContainer(cli *client.Client, id string) error {
+	log.Printf("Removing Container [%s]", id)
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+	return cli.ContainerRemove(ctx, id, types.ContainerRemoveOptions{})
+}
+
+func Write(cli *client.Client, id string, content []string, dest string) error {
```
Comment (thiagomanel) [inline utils/docker_utils.go:135]: what does this write function do?

File: utils/docker_utils.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	"time"
+)
+
+type ContainerConfig struct {
+	Name   string
+	Image  string
+	Mounts []mount.Mount
+}
+
+func NewDockerClient(host string) *client.Client {
```
Comment (thiagomanel) [inline utils/docker_utils.go:41]: log?

File: utils/docker_utils.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	cli, _ := client.NewEnvClient()
+	return cli
+}
+
+func ListContainer(cli *client.Client) {
+	ctns, err := cli.ContainerList(context.Background(), types.ContainerListOptions{All: true})
+	if err != nil {
+		log.Fatal(err)
+	}
+	for _, container := range ctns {
```
Comment (thiagomanel) [inline utils/docker_utils.go]: why are we writing to the stdout? (not sure If understood this correctly, though)

File: utils/docker_utils.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+	for _, c := range content {
+		c = strings.ReplaceAll(c, "'", "'\"'\"'")
+		cmd := fmt.Sprintf(`echo -E '%s' >> %s`, c, dest)
+		log.Printf("Writing [%s] on [%s] from Container [%s]", c, dest, id)
+		Exec(cli, id, cmd)
+	}
+	return nil
+}
+
+func Copy(cli *client.Client, id, src, dest string) error {
```
Comment (thiagomanel) [inline utils/docker_utils.go:159]: what does this function do?

File: utils/docker_utils.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+	}
+	return result
+}
+
+func Pull(cli *client.Client, image string) (io.ReadCloser, error) {
+	reader, err := cli.ImagePull(context.Background(), image, types.ImagePullOptions{})
+	return reader, err
+}
+
+func CheckImage(cli *client.Client, image string) (exist bool, err error) {
```
Comment (thiagomanel) [inline utils/docker_utils.go]: what does this function check?

File: utils/docker_utils.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+		AttachStdout: true,
+		Cmd:          []string{"/bin/bash", "-c", "cat " + path},
+	}
+	rid, err := cli.ContainerExecCreate(context.Background(), id, config)
+	if err != nil {
+		log.Println(err)
+		log.Println("error on creating container exec")
+	}
+	log.Println(rid.ID)
+	hijack, _ := cli.ContainerExecAttach(context.Background(), rid.ID, config)
```
Comment (thiagomanel) [inline utils/docker_utils.go]: what is a *hijack*?

File: utils/docker_utils.go [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+func Copy(cli *client.Client, id, src, dest string) error {
+	log.Printf("Copy [%s] to [%s] from Container [%s]", src, dest, id)
+	dat, _ := ioutil.ReadFile(src)
+	content := string(dat)
+	content = strings.ReplaceAll(content, "'", "'\"'\"'")
+	cmd := fmt.Sprintf("echo -E '%s' >| %s", content, dest)
+	return Exec(cli, id, cmd)
+}
+
+func Exec(cli *client.Client, id, cmd string) error {
```
Comment (thiagomanel) [inline utils/docker_utils.go:177]: I'm wondering about the trade-offs of executing commands one by one.

File: go.sum [hunk 8]
```diff
@@ -1,2 +1,42 @@
+github.com/Microsoft/go-winio v0.4.14 h1:+hMXMk01us9KgxGb7ftKQt2Xpf5hH/yky+TDA+qxleU=
```
Comment (thiagomanel) [inline go.sum:1]: do we really require all these deps? sounds like a NodeJS app 😄

**Review by thiagomanel — CHANGES_REQUESTED** [review]
mostly, questions

File: utils/docker_utils.go (hunk 1 above)
Comment (raonismaneoto) [reply utils/docker_utils.go:135]: Iterates over the content and write each one to the destination file

File: utils/docker_utils.go (hunk 2 above)
Comment (raonismaneoto) [reply utils/docker_utils.go:41]: done

File: utils/docker_utils.go (hunk 3 above)
Comment (raonismaneoto) [reply utils/docker_utils.go]: I can remove this func actually.. gonna do that

File: utils/docker_utils.go (hunk 4 above)
Comment (raonismaneoto) [reply utils/docker_utils.go:159]: it copies the src file (that lives in the host) to the dest (in the container)

File: utils/docker_utils.go (hunk 5 above)
Comment (raonismaneoto) [reply utils/docker_utils.go]: it checks if the image is valid. It must exists on dockerhub.

File: utils/docker_utils.go (hunk 6 above)
Comment (raonismaneoto) [reply utils/docker_utils.go]: it's the way we have to "kidnap" the output of a command executed inside the container..

File: utils/docker_utils.go (hunk 7 above)
Comment (raonismaneoto) [reply utils/docker_utils.go:177]: The task commands you mean? If so, we do exec one by one, but inside the container (scenes of next prs).. The others commands we need to execute (to setup the environment, for example) are executed using this func, but I think it doesn't add any performance issue...

File: go.sum (hunk 8 above)
Comment (raonismaneoto) [reply go.sum:1]: yeah we do. Come on, it's just one file, check some node_modules out there, there is not even comparison hahaha

**Comment by raonismaneoto** [comment]
I've answered/addressed your questions @thiagomanel

File: go.sum (hunk 8 above)
Comment (thiagomanel) [reply go.sum:1]: fair enough

**Review by thiagomanel — CHANGES_REQUESTED** [review]
I have a gut feeling I'll dislike the design of how we coordinate the execution of Tasks into docker layer. I guess we made some decisions that are ok on the previous arrebol but inferior in this new design.

meanwhile, I would love to see docs in all the functions of this docker_util file. the docs should explain the purpose of each function, its parameters and returned values (as well as error conditions). Be free to forward the request to @wesleymonte (I guess he implemented the file, right?)

File: utils/docker_utils.go (hunk 5 above)
Comment (joivo) [reply utils/docker_utils.go]: maybe the image registry can be customizable here

Comment (thiagomanel) [reply utils/docker_utils.go]: Supporting registry other that dockerhub is a very good idea (not sure this  was the suggestion @emanueljoivo made, though)

**Comment by raonismaneoto** [comment]
Yes, it was @wesleymonte who wrote this file, but he also explained it to me. I think I can handle this @thiagomanel

File: utils/docker_utils.go [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+
+//Executes cat in a file inside the container and returns its output
+//Params:
+//cli - the docker client
+//id - the container id
+//path - the file path inside the container
+//It returns:
+//1. nil and an error if the id doesn't exists,
+//or if the file path is invalid.
+//2. The file content as byte array and nil otherwise.
```
Comment (thiagomanel) [inline utils/docker_utils.go:194]: rename this function. Maybe, to `read(cli, id, path) ([]byte, error)`

**Review by thiagomanel — CHANGES_REQUESTED** [review]
I think this docket_util module is more complex than we need. Alternatives design could: 1) use a single file to command the task execution (by wrapping the tasks commands with std and err redirections); 2) map a directory between host and container to stage in/out scripts and output). However, I'm ok with following your design. 

Anyway, I added a new request to rename a function. Once you changed it, I'm ok to accept the PR

File: utils/docker_utils.go (hunk 9 above)
Comment (raonismaneoto) [reply utils/docker_utils.go:194]: this function actually reads the content that comes from a connection (a socket). See Cat func, it uses the aux read func to read the content of a hijack connection. In short, the cat executes the unix `cat` inside the container and expose the result to the hijack connection, then, the read func get this content.

**Comment by raonismaneoto** [comment]
> I think this docket_util module is more complex than we need. Alternatives design could: 1) use a single file to command the task execution (by wrapping the tasks commands with std and err redirections); 2) map a directory between host and container to stage in/out scripts and output). However, I'm ok with following your design.
> 
> Anyway, I added a new request to rename a function. Once you changed it, I'm ok to accept the PR

I agree with you, but I need to sleep on that to figure out a issue description for this. By now, I think this discussion and the notes I made are enough to keep this in our mind.

File: utils/docker_utils.go (hunk 9 above)
Comment (thiagomanel) [reply utils/docker_utils.go:194]: still not convinced. rename to `read` as I suggested. `cat` means `concat`.

Comment (raonismaneoto) [reply utils/docker_utils.go:194]: done

**Review by thiagomanel — APPROVED** [review]
lgtm

**Pull request merged**